dali/test/python/test_external_source_parallel_utils.py

```diff
@@ -0,0 +1,155 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+import nvidia.dali as dali
+
+from test_utils import compare_pipelines, check_batch, RandomDataIterator, RandomlyShapedDataIterator
+
+
+class ExtCallback:
+    """Callable to generate specified data samples"""
+
+    def __init__(
+            self, dims, epoch_size, dtype, exception_class=StopIteration, random_data=False,
+            random_shape=False):
+        self.dims = dims
+        self.epoch_size = epoch_size
+        self.dtype = dtype
+        self.exception_class = exception_class
+        self.ds = {}
+        self.random_data = random_data
+        self.random_shape = random_shape
+        self.data_iterator = None
+        self.iterator_data_samples = []
+        if random_data and not random_shape:
+            self.data_iterator = iter(RandomDataIterator(1, shape=dims, dtype=dtype))
+        if random_data and random_shape:
+            self.data_iterator = iter(RandomlyShapedDataIterator(1, max_shape=dims, dtype=dtype))
+        if not random_data and random_shape:
+            raise ValueError("If random_shape is required the random_data is required to be True.")
+
+    def __call__(self, sample_info):
+        if sample_info.idx_in_epoch >= self.epoch_size:
+            raise self.exception_class
+        if self.data_iterator:
+            while len(self.iterator_data_samples) <= sample_info.idx_in_epoch:
+                batch = self.data_iterator.next()
+                self.iterator_data_samples.append(batch[0])
+            return self.iterator_data_samples[sample_info.idx_in_epoch]
+        if sample_info.idx_in_epoch not in self.ds:
+            self.ds[sample_info.idx_in_epoch] = np.full(
+                self.dims, sample_info.idx_in_epoch, dtype=self.dtype)
+        return self.ds[sample_info.idx_in_epoch]
+
+
+class ExtCallbackTensorCPU(ExtCallback):
+
+    def __call__(self, sample_info):
+        return dali.tensors.TensorCPU(super().__call__(sample_info))
+
+
+def create_pipe(
+        callback, device, batch_size, num_outputs=None, layout=None, py_num_workers=None,
+        py_start_method="fork", parallel=True, device_id=0):
+    pipe = dali.pipeline.Pipeline(
+        batch_size, 1, device_id, py_num_workers=py_num_workers, py_start_method=py_start_method)
+    inputs = dali.fn.external_source(
+        callback, num_outputs=num_outputs, device=device, layout=layout, batch=False,
+        parallel=parallel)
+    if num_outputs is None:
+        pipe.set_outputs(inputs)
```

Comment:
We really should hide this logic inside `set_outputs`

Reply:
Out of scope. And this is already an API, that returns the output in default case or list of outputs if there is `num_outputs` provided.

Reply:
Just a thought....